### PR TITLE
[python-package] Allow to pass Arrow table and array as init scores

### PR DESCRIPTION
**Pull request by borchero**
# Motivation

This is part of a series of PRs towards #6022 and depends on #6166.

For a legible diff, see https://github.com/borchero/LightGBM/compare/arrow-support-groups...arrow-support-init-scores.

### CODE REVIEW

File: tests/python_package_test/test_arrow.py
Comment (jameslamb) [inline tests/python_package_test/test_arrow.py:255]: Can you please add stricter checks here than just shape and `dtype`? Specifically, I'd like to see that the input values are correctly preserved, in the correct order.

I don't see anything in these tests that's checking that the order of the input is correctly preserved for multi-class classification (where a multi-dimensional `init_score` is provided). And that's definitely an area where bugs could be found, as it requires knowing where LightGBM requires row-major vs. column-major ordering of data.

File: python-package/lightgbm/basic.py
Comment (jameslamb) [inline python-package/lightgbm/basic.py:2884]: Why is this `(for multi-class task)` qualifier necessary?

Couldn't all of a `pyarrow.Array`, `pyarrow.ChunkedArray`, and `pyarrow.Table` be used to provide a 2-dimensional matrix? Notice that we don't have `(for multi-class task)` in the docstring entry for `numpy array`, for exactly that reason.

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks for this! I support it overall.

I think it should be possible to provide 2D `pyarrow.Array` and `pyarrow.ChunkedArray` objects for `init_score` as well. Are there some limitations there that I'm missing, that led you to only designate `pyarrow.Table` as "for multi-class" in the docs + not add tests on 2D `pyarrow.Array` / `pyarrow.ChunkedArray`?

If that was just an oversight, please add some tests on providing 2D data for `init_score` via the `pyarrow` array types.

**Comment by borchero** [comment]
> I think it should be possible to provide 2D pyarrow.Array and pyarrow.ChunkedArray objects

I'm not sure I mistook something in the Arrow design but aren't Arrow arrays 1D by design? In my mind, there's no n-d Arrow arrays (like you have NumPy arrays).

**Comment by jameslamb** [comment]
> aren't Arrow arrays 1D by design

Oh I guess they are: https://blog.djnavarro.net/posts/2022-05-25_arrays-and-tables-in-arrow/#data-objects

(the author was working at Voltron on Arrow when that was written, I believe)

Sorry, I'm not that familiar with Arrow... trying to get caught up to review these PRs. I think this was the first of these PRs that included both the Array types and Table, so the first place I really thought about it (e.g. the absence of `Array` never occurred to me in #6034).

Thanks for your patience and quick answers.

File: python-package/lightgbm/basic.py
Comment (jameslamb) [reply python-package/lightgbm/basic.py:2884]: dismissing this, per https://github.com/microsoft/LightGBM/pull/6167#issuecomment-1823611526

File: python-package/lightgbm/basic.py
Comment (borchero) [reply python-package/lightgbm/basic.py:2589]: A bunch of issues due to the generality of `data` and the absence of proper typing form the `pyarrow` package:

```
python-package/lightgbm/basic.py:2589: error: Item "list[list[float]]" of "list[list[float]] | Any | list[list[int]] | list[float] | list[int]" has no attribute "columns"  [union-attr]
python-package/lightgbm/basic.py:2589: error: Item "list[list[int]]" of "list[list[float]] | Any | list[list[int]] | list[float] | list[int]" has no attribute "columns"  [union-attr]
python-package/lightgbm/basic.py:2589: error: Item "list[float]" of "list[list[float]] | Any | list[list[int]] | list[float] | list[int]" has no attribute "columns"  [union-attr]
python-package/lightgbm/basic.py:2589: error: Item "list[int]" of "list[list[float]] | Any | list[list[int]] | list[float] | list[int]" has no attribute "columns"  [union-attr]
```

File: tests/python_package_test/test_arrow.py
Comment (borchero) [reply tests/python_package_test/test_arrow.py]: I think the latest adjustment to the test case should cover this now.

File: python-package/lightgbm/basic.py
Comment (jameslamb) [reply python-package/lightgbm/basic.py:2589]: Ok got it, thanks! I think it's ok to suppress that warning for now then.

**Review by jameslamb — APPROVED** [review]
Looks great to me, thanks!

@shiyu1994 or @guolinke could you help with a review as well?

File: tests/python_package_test/test_arrow.py
Comment (jameslamb) [reply tests/python_package_test/test_arrow.py]: yep that adjustment looks good to me, thanks!

**Review by guolinke — APPROVED** [review]
LGTM

**Comment by borchero** [comment]
Can we continue here @jameslamb? 🙏🏼 would love to get Arrow support in a LightGBM release before Christmas :smile:

**Comment by jameslamb** [comment]
I dont want to merge this until #6216 is merged, as that fixes an issue causing CI to fail deterministically. That is just waiting for a review. Sorry for the delay.

**Comment by jameslamb** [comment]
Ok #6216 is merged, thanks to @guolinke !

And thanks to @shiyu1994 , I could merge `master` into this by clicking a button on my phone 😁 

Once CI passes, I'll merge this and we can move on to the next one. Thanks as always for your patience and help @borchero 😊

**Comment by borchero** [comment]
What's up with the CUDA CI, is this a problem across this repo? :eyes:

**Comment by jameslamb** [comment]
> is this a problem across the repo?

It seems like it, yes.

You can see all runs of GitHub Actions in a project at the `Actions` tab. Here are the recent runs of CUDA across all commits:

https://github.com/microsoft/LightGBM/actions/workflows/cuda.yml

Looks to me like many of them failed yesterday (around 15-20 hours ago) with the same types of errors.

> The job running on runner nv6-01 has exceeded the maximum execution time of 60 minutes.
> The operation was canceled.

([example build not from this PR](https://github.com/microsoft/LightGBM/actions/runs/7055508228))

These run on a dedicated VM Microsoft pays for, not GitHub's fleet of hosted runners, so there can only be one job running at a time across the entire project. That makes it easy for them to get backed up.

I do see a task got picked up and is running right now for another PR: https://github.com/microsoft/LightGBM/actions/runs/7061296005/job/19222636772

Let's see if that succeeds. Sorry for the inconvenience.

**Comment by jameslamb** [comment]
Yeah the CUDA CI runs are all timing out at this point:

https://github.com/microsoft/LightGBM/blob/5083df15c6866a78704787ca942931feaa096a76/.github/workflows/cuda.yml#L106

@shiyu1994 can you please look into it? Maybe try shelling into that box and running the steps from the CI configuration manually, to see if you can get more logs?

I suspect it might be something like an interactive dialog that's being triggered, and then timing out waiting for that.

**Comment by jameslamb** [comment]
@shiyu1994 can you please help with the CUDA machine or give me access so I can do it?

The most recent manually-triggered re-run for this PR has been stuck for almost 24 hours.

image

> Requested labels: self-hosted, linux
Job defined at: microsoft/LightGBM/.github/workflows/cuda.yml@refs/pull/6167/merge
Waiting for a runner to pick up this job...

([build link](https://github.com/microsoft/LightGBM/actions/runs/7054883149/job/19252134691?pr=6167))

Thank you.

**Comment by jameslamb** [comment]
Thanks so much @shiyu1994 ! I just restarted the builds for this PR.

**Comment by jameslamb** [comment]
🎉 that worked, thank you again!

https://github.com/microsoft/LightGBM/actions/runs/7054883149?pr=6167

**Pull request merged**
